### PR TITLE
Add fetching guilds over rest

### PR DESCRIPTION
**Pull request by Plerx2493**
# Summary
Adds the `/users/@me/guilds` endpoint to the DiscordApiClient and exposes a `IAsyncEnumerable<DiscordGuild>` through `DiscordClient#GetGuilds*Async`

### CODE REVIEW

**Comment by Plerx2493** [comment]
Did some basic testing. Pagination isnt tested because i have no bot in more than 200 guilds but the code is taken from `GetMessagesAsync`

File: DSharpPlus/Clients/DiscordClient.cs [hunk 1]
```diff
@@ -835,6 +835,91 @@ public async Task<DiscordApplicationCommand> EditGuildApplicationCommandAsync(ul
...
+        int lastCount;
+        do
+        {
+            int fetchSize = remaining > 100 ? 100 : remaining;
+            IReadOnlyList<DiscordGuild> fetchedGuilds = await this.ApiClient.GetGuildsAsync( fetchSize, isbefore ? last ?? before : null, !isbefore ? last ?? after : null, withCount);
+
+            lastCount = fetchedGuilds.Count;
+            remaining -= lastCount;
+            
+            //We sort the returned guilds by ID so that they are in order in case Discord switches the order AGAIN.
```
Comment (akiraveliara) [inline DSharpPlus/Clients/DiscordClient.cs:898]: are we sure this is needed

File: DSharpPlus/Clients/DiscordClient.cs [hunk 2]
```diff
@@ -835,6 +835,91 @@ public async Task<DiscordApplicationCommand> EditGuildApplicationCommandAsync(ul
...
+            DiscordGuild[] sortedGuildsArray = fetchedGuilds.ToArray();
+            Array.Sort(sortedGuildsArray, (x, y) => x.Id.CompareTo(y.Id));
+            
+
+            if (!isbefore)
+            {
+                foreach (DiscordGuild msg in sortedGuildsArray)
+                {
+                    yield return msg;
```
Comment (akiraveliara) [inline DSharpPlus/Clients/DiscordClient.cs:907]: naming

File: DSharpPlus/Clients/DiscordClient.cs [hunk 3]
```diff
@@ -835,6 +835,91 @@ public async Task<DiscordApplicationCommand> EditGuildApplicationCommandAsync(ul
...
+    /// </summary>
+    /// <exception cref="BadRequestException">Thrown when an invalid parameter was provided.</exception>
+    /// <exception cref="ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+    public IAsyncEnumerable<DiscordGuild> GetGuildsBeforeAsync(ulong before, int limit = 200, bool? withCount = null)
+        => this.GetGuildsInternalAsync(limit, before, null, null);
+
+    /// <summary>
+    /// Returns a list of guilds after a certain guild.
+    /// <param name="limit">The amount of guilds to fetch.</param>
+    /// <param name="after">guild to fetch after from.</param>
```
Comment (akiraveliara) [inline DSharpPlus/Clients/DiscordClient.cs:854]: incomplete sentence

**Review by akiraveliara — COMMENTED** [review]
few nitpicks, also; is `withCount` the intentional parameter naming? there's two counts being controlled by that parameter, so

File: DSharpPlus/Clients/DiscordClient.cs (hunk 1 above)
Comment (Plerx2493) [reply DSharpPlus/Clients/DiscordClient.cs:898]: Im not really sure. They switched the order on the messages endpoint sometimes

**Pull request merged**